### PR TITLE
188 update chart tooltip

### LINKED ISSUE #188: Update chart tooltip
For the chart tooltip...

Display each field label along with the value and the target. It should look like

```
[SITE NAME]

[VAL A] - [VALUE] / [TARGET]
[VAL B] - [VALUE] / [TARGET]
...

TOTAL - [SUM OF VALUES] / [SUM OF TARGETS]
```

Do not display the "N/A" group unless it has been configured as a value/label pair when the chart was created/updated.

Image

### PR DESCRIPTION
**Pull request by Elkrival**
Made updates to the tooltip content and ui display.

Screen Shot 2022-09-29 at 1 43 48 PM

### CODE REVIEW

File: views/fe-utils/siteTooltipContent.js [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+        const rowValue = { labelColumn: valueName, valueColumn }
+        siteTotalCount = siteTotalCount + payload.counts[valueName]
+        siteTotalPercent = siteTotalPercent + +valuePercent
+        tooltipData.push(rowValue)
+      }
+    })
+
+  tooltipData.push({
+    labelColumn: totalLabel,
+    valueColumn: `${siteTotalCount}/(${siteTotalPercent}%)`,
```
Comment (mikestone14) [inline views/fe-utils/siteTooltipContent.js]: I think we should use our percent formatter for this and the other tooltip rows.

File: views/fe-utils/siteTooltipContent.js [hunk 2]
```diff
@@ -0,0 +1,27 @@
+export const siteTooltipContent = (siteData) => {
+  const tooltipData = []
+  const reserved = 'N/A'
+  const totalLabel = 'Total'
```
Comment (mikestone14) [inline views/fe-utils/siteTooltipContent.js]: can we have these live in front-end constants somewhere so we can reference them where needed throughout the front-end?

File: views/styles/chart_styles.js [hunk 3]
```diff
@@ -100,4 +100,38 @@ export const chartStyles = (theme) => ({
...
+    width: '14vw',
+    padding: '1vw 1vw .5vw 1vw ',
+    background: 'rgba(255, 255, 255, 0.9)',
+  },
+  tooltipHeaderRow: {
+    display: 'flex',
+    flexDirection: 'row',
+    flexWrap: 'wrap',
+    width: '100%',
+    paddingBottom: '2%',
```
Comment (mikestone14) [inline views/styles/chart_styles.js]: where are these values coming from? I'm noticing we're using a few different units like %, px, and vw. Can we consolidate to use px or rem?

File: views/styles/chart_styles.js [hunk 4]
```diff
@@ -100,4 +100,38 @@ export const chartStyles = (theme) => ({
...
+    flexWrap: 'wrap',
+    width: '100%',
+    paddingTop: '2%',
+    paddingBottom: '2%',
+  },
+  tooltipLabelColumn: {
+    display: 'flex',
+    flexDirection: 'column',
+    flexBasis: '40%',
+    flex: '1',
```
Comment (mikestone14) [inline views/styles/chart_styles.js]: if we make this `flex: '2'` and the value column `flex: '3'` can we remove the `flexBasis`?

File: views/styles/chart_styles.js [hunk 5]
```diff
@@ -100,4 +100,38 @@ export const chartStyles = (theme) => ({
...
+    display: 'flex',
+    flexDirection: 'row',
+    flexWrap: 'wrap',
+    width: '100%',
+    paddingBottom: '2%',
+    borderBottom: 'solid gray 1px',
+  },
+  tooltipContentRow: {
+    display: 'flex',
+    flexDirection: 'row',
```
Comment (mikestone14) [inline views/styles/chart_styles.js]: this is the default direction so you don't need to specify it. Also, do we need this to be flex?

File: views/styles/chart_styles.js [hunk 6]
```diff
@@ -100,4 +100,38 @@ export const chartStyles = (theme) => ({
   publicText: {
     alignSelf: 'center',
   },
+  tooltipContainer: {
+    width: '14vw',
+    padding: '1vw 1vw .5vw 1vw ',
+    background: 'rgba(255, 255, 255, 0.9)',
```
Comment (mikestone14) [inline views/styles/chart_styles.js]: can we use our color constants?

File: views/components/Graphs/BarGraphTooltip.jsx [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+              <div className={classes.tooltipLabelColumn}>{labelColumn}</div>
+              <div className={classes.tooltipValueColumn}>{valueColumn}</div>
+            </div>
+          )
+        })}
+      </div>
+    )
+  }
+
+  return null
```
Comment (mikestone14) [inline views/components/Graphs/BarGraphTooltip.jsx]: I think it's common to return null first if conditions are not met and otherwise return the component content. So in this case it would be something like...

```jsx
const BarGraphTooltip = ({ classes, active, payload, label }) => {
  if (!active || !payload?.length) {
    return null
  }

  return (
    ...
  )
}
```

Comment (Elkrival) [reply views/components/Graphs/BarGraphTooltip.jsx]: https://github.com/AMP-SCZ/dpdash/pull/191/commits/82a8267770a707eedb1efe2f1cc9193d667b1d12

File: views/styles/chart_styles.js (hunk 6 above)
Comment (Elkrival) [reply views/styles/chart_styles.js]: https://github.com/AMP-SCZ/dpdash/pull/191/commits/82a8267770a707eedb1efe2f1cc9193d667b1d12

File: views/styles/chart_styles.js (hunk 5 above)
Comment (Elkrival) [reply views/styles/chart_styles.js]: https://github.com/AMP-SCZ/dpdash/pull/191/commits/82a8267770a707eedb1efe2f1cc9193d667b1d12

Yes otherwise the row stack the columns vertically

File: views/styles/chart_styles.js (hunk 4 above)
Comment (Elkrival) [reply views/styles/chart_styles.js]: https://github.com/AMP-SCZ/dpdash/pull/191/commits/82a8267770a707eedb1efe2f1cc9193d667b1d12

File: views/styles/chart_styles.js (hunk 3 above)
Comment (Elkrival) [reply views/styles/chart_styles.js]: https://github.com/AMP-SCZ/dpdash/pull/191/commits/82a8267770a707eedb1efe2f1cc9193d667b1d12

File: views/fe-utils/siteTooltipContent.js (hunk 2 above)
Comment (Elkrival) [reply views/fe-utils/siteTooltipContent.js]: https://github.com/AMP-SCZ/dpdash/pull/191/commits/82a8267770a707eedb1efe2f1cc9193d667b1d12

File: views/fe-utils/siteTooltipContent.js (hunk 1 above)
Comment (Elkrival) [reply views/fe-utils/siteTooltipContent.js]: https://github.com/AMP-SCZ/dpdash/pull/191/commits/82a8267770a707eedb1efe2f1cc9193d667b1d12

File: views/fe-utils/siteTooltipContent.js [hunk 8]
```diff
@@ -0,0 +1,30 @@
+import { TOTAL_LABEL, NA } from '../../constants/vars'
+import { formatAsPercentage } from './formatAsPercentageUtil'
+
+export const siteTooltipContent = (siteData) => {
+  const tooltipData = []
+  let siteTotalCount = 0
+  let siteTotalPercent = 0
```
Comment (mikestone14) [inline views/fe-utils/siteTooltipContent.js]: do we not get these values from the `studyTotals` or some other key in the site data?

File: views/fe-utils/siteTooltipContent.js [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+
+  siteData
+    .filter(({ name }) => name !== NA)
+    .forEach(({ name: valueName, payload }) => {
+      if (valueName !== NA) {
+        const valuePercent = payload.percentages[valueName]
+        const valueColumn = `${payload.counts[valueName]}/(${formatAsPercentage(
+          valuePercent
+        )})`
```
Comment (mikestone14) [inline views/fe-utils/siteTooltipContent.js]: this should be `COUNT / TARGET (PERCENT%)`. Right now it looks like `COUNT/(PERCENT%)`

File: views/fe-utils/siteTooltipContent.js [hunk 10]
```diff
@@ -0,0 +1,30 @@
+import { TOTAL_LABEL, NA } from '../../constants/vars'
```
Comment (mikestone14) [inline views/fe-utils/siteTooltipContent.js]: should this be an `index` file within `constants`? These are all going to be variables that get exported.

File: constants/vars.js [hunk 11]
```diff
@@ -0,0 +1,2 @@
+export const NA = 'N/A'
```
Comment (mikestone14) [inline constants/vars.js]:
```suggestion
export const N_A = 'N/A'
```

File: views/fe-utils/siteTooltipContent.js [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+import { formatAsPercentage } from './formatAsPercentageUtil'
+
+export const siteTooltipContent = (siteData) => {
+  const tooltipData = []
+  let siteTotalCount = 0
+  let siteTotalPercent = 0
+
+  siteData
+    .filter(({ name }) => name !== NA)
```
Comment (mikestone14) [inline views/fe-utils/siteTooltipContent.js]: let's update the other places where we use the `N_A` and `TOTALS` constants as well.

File: views/fe-utils/siteTooltipContent.js (hunk 9 above)
Comment (Elkrival) [reply views/fe-utils/siteTooltipContent.js]: https://github.com/AMP-SCZ/dpdash/pull/191/commits/4a43fd4984d56defdb4112250a883f89a70016a2

File: views/fe-utils/siteTooltipContent.js (hunk 8 above)
Comment (Elkrival) [reply views/fe-utils/siteTooltipContent.js]: https://github.com/AMP-SCZ/dpdash/pull/191/commits/4a43fd4984d56defdb4112250a883f89a70016a2

File: views/fe-utils/siteTooltipContent.js (hunk 10 above)
Comment (Elkrival) [reply views/fe-utils/siteTooltipContent.js]: https://github.com/AMP-SCZ/dpdash/pull/191/commits/830a204d6e575727759e922645ae04fe835fda58

File: constants/vars.js (hunk 11 above)
Comment (Elkrival) [reply constants/vars.js]: https://github.com/AMP-SCZ/dpdash/pull/191/commits/830a204d6e575727759e922645ae04fe835fda58

File: views/fe-utils/siteTooltipContent.js (hunk 12 above)
Comment (Elkrival) [reply views/fe-utils/siteTooltipContent.js]: https://github.com/AMP-SCZ/dpdash/pull/191/commits/32872dc7c65af9e91a8eaeffc68e80a7ca35bd2f

https://github.com/AMP-SCZ/dpdash/pull/191/commits/e9d38f2a91f03adbc64f06a73491bac4ad230b82

File: server/controllers/chartsController.js [hunk 13]
```diff
@@ -33,6 +33,10 @@ const postProcessData = (data, studyTotals) => {
           ...existingEntriesForStudy.percentages,
           [valueLabel]: percent,
         },
+        targets: {
```
Comment (mikestone14) [inline server/controllers/chartsController.js:36]: what's the reason for this change? I thought we already had access to the study totals/targets?

File: views/components/Graphs/BarGraphTooltip.jsx [hunk 14]
```diff
@@ -0,0 +1,35 @@
+import React from 'react'
+import { TOTAL_LABEL } from '../../../constants'
+import { siteTooltipContent } from '../../fe-utils/siteTooltipContent'
+
+const BarGraphTooltip = ({ classes, active, payload, label, studyTotals }) => {
+  if (!active || !payload?.length) {
+    return null
+  }
+  return (
```
Comment (mikestone14) [inline views/components/Graphs/BarGraphTooltip.jsx]:
```suggestion
  }

  return (
```

File: views/fe-utils/siteTooltipContent.js [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+  let siteTotalPercent = 0
+
+  siteData
+    .filter(({ name }) => name !== N_A)
+    .forEach(({ name: valueName, payload }) => {
+      if (valueName !== N_A) {
+        const valuePercent = payload.percentages[valueName]
+        const valueColumn = `${payload.counts[valueName]}/${
+          payload.targets[valueName]
+        }(${formatAsPercentage(valuePercent)})`
```
Comment (mikestone14) [inline views/fe-utils/siteTooltipContent.js]:
```suggestion
        } (${formatAsPercentage(valuePercent)})`
```

File: views/fe-utils/siteTooltipContent.js [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+        siteTotalPercent = siteTotalPercent + +valuePercent
+        tooltipData.push(rowValue)
+      }
+    })
+
+  tooltipData.push({
+    labelColumn: TOTAL_LABEL,
+    valueColumn: `${siteTotals.count}/${
+      siteTotals?.targetTotal || 0
+    }(${formatAsPercentage(siteTotalPercent)})`,
```
Comment (mikestone14) [inline views/fe-utils/siteTooltipContent.js]:
```suggestion
    } (${formatAsPercentage(siteTotalPercent)})`,
```

File: views/fe-utils/siteTooltipContent.js [hunk 17]
```diff
@@ -0,0 +1,30 @@
+import { N_A, TOTAL_LABEL } from '../../constants'
+import { formatAsPercentage } from './formatAsPercentageUtil'
+
+export const siteTooltipContent = (siteData, siteTotals) => {
+  const tooltipData = []
+  let siteTotalPercent = 0
```
Comment (mikestone14) [inline views/fe-utils/siteTooltipContent.js]: don't we have this already in the site data or tooltip payload?

File: views/fe-utils/siteTooltipContent.js (hunk 9 above)
Comment (mikestone14) [reply views/fe-utils/siteTooltipContent.js]: now it looks like `COUNT/TARGET(PERCENT%)` not `COUNT / TARGET (PERCENT%)`

File: views/fe-utils/siteTooltipContent.js (hunk 8 above)
Comment (mikestone14) [reply views/fe-utils/siteTooltipContent.js]: it looks like there is still a variable that we can get from study totals.

File: views/styles/chart_styles.js [hunk 18]
```diff
@@ -118,4 +118,29 @@ export const chartStyles = (theme) => ({
...
     outlineOffset: 'none',
   },
+  tooltipContainer: {
+    width: '200px',
+    padding: '14px 14px 7px 14px ',
+    background: colors.transparentWhite,
+  },
+  tooltipHeaderRow: {
+    display: 'flex',
+    width: '200px',
```
Comment (mikestone14) [inline views/styles/chart_styles.js]: can this be 100%?

File: views/styles/chart_styles.js [hunk 19]
```diff
@@ -118,4 +118,29 @@ export const chartStyles = (theme) => ({
...
+    paddingBottom: '4px',
+    borderBottom: 'solid gray 0.5px',
+  },
+  tooltipContentRow: {
+    display: 'flex',
+    width: '200px',
```
Comment (mikestone14) [inline views/styles/chart_styles.js]: can this be 100%?

File: server/controllers/chartsController.js (hunk 13 above)
Comment (Elkrival) [reply server/controllers/chartsController.js:36]: There's a studyTotals object and that has the summation of the sites targets, this change adds the individual target of the value so that when we display a row of a value we can plug in the individual target to match the Value/Target(%) piece. It is currently missing from the charts data object and the above is the reason why it was added.

File: views/fe-utils/siteTooltipContent.js (hunk 17 above)
Comment (Elkrival) [reply views/fe-utils/siteTooltipContent.js]: The data we have is the N/A percentage but that's the difference of the sites current targets. In the ticket I think we want the summation of the percentages. Maybe I'm missing something?
Screen Shot 2022-10-03 at 3 17 44 PM

File: server/controllers/chartsController.js (hunk 13 above)
Comment (Elkrival) [reply server/controllers/chartsController.js:36]: Screen Shot 2022-10-03 at 3 17 54 PM

File: views/fe-utils/siteTooltipContent.js (hunk 17 above)
Comment (mikestone14) [reply views/fe-utils/siteTooltipContent.js]: don't we have a site total count and a site total target? If so I believe we could calculate the percentage from those values.

File: views/styles/chart_styles.js (hunk 19 above)
Comment (Elkrival) [reply views/styles/chart_styles.js]: https://github.com/AMP-SCZ/dpdash/pull/191/commits/4ed7b45e0dfa83017cab82c474a3512ffc0c38c2

File: views/styles/chart_styles.js (hunk 18 above)
Comment (Elkrival) [reply views/styles/chart_styles.js]: https://github.com/AMP-SCZ/dpdash/pull/191/commits/4ed7b45e0dfa83017cab82c474a3512ffc0c38c2

File: views/fe-utils/siteTooltipContent.js (hunk 16 above)
Comment (Elkrival) [reply views/fe-utils/siteTooltipContent.js]: https://github.com/AMP-SCZ/dpdash/pull/191/commits/4ed7b45e0dfa83017cab82c474a3512ffc0c38c2

File: views/fe-utils/siteTooltipContent.js (hunk 15 above)
Comment (Elkrival) [reply views/fe-utils/siteTooltipContent.js]: https://github.com/AMP-SCZ/dpdash/pull/191/commits/4ed7b45e0dfa83017cab82c474a3512ffc0c38c2

File: views/components/Graphs/BarGraphTooltip.jsx (hunk 14 above)
Comment (Elkrival) [reply views/components/Graphs/BarGraphTooltip.jsx]: https://github.com/AMP-SCZ/dpdash/pull/191/commits/4ed7b45e0dfa83017cab82c474a3512ffc0c38c2

**Review by mikestone14 — APPROVED** [review]
